wg-cloud-provider/cloud-provider-requirements.md

```diff
@@ -0,0 +1,87 @@
+
+* Cloud Controller Manager implementations.
+* Documentation.
+
+This document may be amended with additional locations that relate to enabling
+consistent upstream testing, independent storage drivers, and other code with
+common integration hooks may be added
+
+## Motivation
+
+The development of the Cloud Controller Manager[3] and Cloud Provider
+Interface[4] has enabled the provider SIGs to develop external providers that
+capture the core functionality of the upstream providers. By defining the
+expected locations and naming conventions of where the external provider code
+is, we continue in creating a consistent experience for:
+
+* Users of the providers, who will have easily understandable conventions for
+  discovering and using all of the providers.
+* SIG-Docs, who will have a common hook for building or linking to externally
+  managed documentation
+* SIG-Testing, who will be able to use common entry points for enabling
+  provider-specific e2e testing.
+* Future cloud provider authors, who will have a common framework and examples
+  from which to build and share their code base.
+
+## Requirements
+
+Each cloud provider hosted within the `kubernetes` organization shall have a
+single repository named `kubernetes/cloud-provider-<provider_name>`. Those
+repositories shall have the following structure:
+
+* A `cloud-controller-manager` subdirectory that contains the implementation
+  of the provider-specific cloud controller.
+* A `docs` subdirectory.
+* A `docs/cloud-controller-manager.md` file that describes the options and
+  usage of the cloud controller manager code.
+* A `tests` subdirectory that contains testing code.
+
+Additionally, the repository should have:
+
+* A `docs/getting-started.md` file that describes the installation and basic
+  operation of the cloud controller manager code.
+
+Where the provider has additional capabilities, the repository should have
+the following subdirectories that contain the common features:
+
+* `dns` for DNS provider code.
+* `cni` for the Container Network Interface (CNI) driver.
+* `flex` for the Flex Volume driver.
+* `installer` for custom installer code.
+
+Each repository may have additional directories and files that are used for
+additional feature that include but are not limited to:
+
+* Other provider specific testing.
+* Additional documentation, including examples and developer documentation.
+* Dependencies on provider-hosted or other external code.
+
+## Timeline
+
+To facilitate community development, providers named in the `Make SIGs
+responsible for implementations of CloudProvider` patch[1] can immediately
+migrate their external provider work into their named repositories.
+
+Each provider will work to implement the required structure during the
+Kubernetes 1.11 development cycle, with conformance by the 1.11 release.
+
+After the 1.11 release all current and new provider implementations must
+conform with the requirements outlined in this document.
```

Comment:
I think it's reasonable to assume that the repo structure we want may change (a lot) as we test it out. Having said that we should include in the timeline section that at any time before the 1.11 release we may make drastic changes.

Reply:
I agree with @andrewsykim. This LGTM as it is right now. We still need to add the `csi` sub-dir, but let's get this merged and then edit this as needed.